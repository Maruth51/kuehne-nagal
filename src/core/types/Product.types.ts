
export type TProductCategories = 
  'Software Development' |
  'Daily Business' |
  'Daily Business' |
  'Graphic Editors' |
  'Graphic Editors' |
  'Text Editors' |
  'Management Tools'

  export interface IProduct {
    productName: string;
    tags: string[];
    category: TProductCategories;
    manufacturerUrl: string;
    description: string[];
    option1: string | null;
    option2: string | null;
  }