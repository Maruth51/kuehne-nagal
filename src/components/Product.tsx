import React from "react";
import { IProduct } from "core/types/Product.types";
import styles from 'styles/Product.module.scss';

const Product: React.FC<{
  product: IProduct,
  onSelect: () => void
}> = ({
  product,
  onSelect
}) => {
  return <div
      className={styles.product}
      onClick={onSelect}
      >
        <div className={styles.left}>
          {product.productName}
          <div className={styles.tags}>
            {product.tags.map((tag, idx) => <span key={idx} className={styles.tag}>{tag}</span>)}
          </div>
        </div>
        <div className={styles.right}>
          {product.category}
        </div>
      </div>
}

export default Product;