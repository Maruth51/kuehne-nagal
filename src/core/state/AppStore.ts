import { fetchProducts } from "core/services/productDataService";
import { IProduct, TProductCategories } from "core/types/Product.types";
import { makeAutoObservable } from "mobx";


class AppStore {
  activeTab = 'product'
  selectedCategories: TProductCategories[] = []
  searchString = '';
  selectedProduct: IProduct | null = null;
  products: IProduct[] = []
  constructor() {
    makeAutoObservable(this)
  }

  search(s: string) {
    this.searchString = s
  }

  selectProduct(product: IProduct) {
    this.selectedProduct = product
  }

  get filteredProducts(): IProduct[] {
    if(this.searchString === '') {
      return []
    }
    return this.products.filter((product : IProduct) => {
      // if not category selected, by default all category included
      let isProductCategorySelected = true;
      if(!!this.selectedCategories.length) {
        isProductCategorySelected = this.selectedCategories.includes(product.category)
      }
      return isProductCategorySelected && product.productName.toLowerCase().includes(this.searchString.trim().toLowerCase())
    })
  }

  toggleSelectedCategory(category: TProductCategories) {
    const categoryIdx = this.selectedCategories.indexOf(category)
    if (categoryIdx === -1) {
      this.selectedCategories.push(category)
    } else {
      this.selectedCategories.splice(categoryIdx, 1)
    }
  }

  async fetchProducts() {
    this.products = await fetchProducts()
  }
}

export default AppStore;