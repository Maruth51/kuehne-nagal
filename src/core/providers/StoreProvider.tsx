import React from "react";
import AppStore from "core/state/AppStore";

const StoreContext = React.createContext<AppStore>(new AppStore())

export const useStore = () => React.useContext(StoreContext);

export const StoreProvider: React.FC<{ children: React.ReactNode, store: AppStore}> = ({ children, store }) => {
  return (
    <StoreContext.Provider value={store}>{children}</StoreContext.Provider>
  );
};