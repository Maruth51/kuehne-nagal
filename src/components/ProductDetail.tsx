import React from "react";
import { useStore } from "core/providers/StoreProvider";
import { observer } from "mobx-react-lite";
import styles from "styles/ProductDetail.module.scss"

const ProductDetail = () => {
  const { selectedProduct } = useStore()
  return <div data-testid='product-detail' className={styles.container}>
    {selectedProduct && <>
      <div className={styles.head}>Product Details</div>
      <div className={styles.productDetail}>
        <span>{selectedProduct.productName} </span>
        <div className={styles.tags}>
          {selectedProduct.tags.map((tag, idx) => <span key={idx} className={styles.tag}>{tag}</span>)}
        </div>
        <a type="button" href={selectedProduct.manufacturerUrl} target="_blank" rel='noreferrer' className={styles.actionBtn}>Go to Manufacturer</a>
        <p className={styles.description}>
          {selectedProduct.description[0]}
        </p>
        {selectedProduct.option1 && <div className="option">
          <input type="radio" name="maintanance-option" id='option1'></input>
          <label htmlFor="option1"> {selectedProduct.option1}</label>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
          </p>
        </div>}
        {selectedProduct.option2 && <div className="option">
          <p>
            <input type="radio" name="maintanance-option" id='option2'></input>
            <label htmlFor="option2"> {selectedProduct.option2}</label>
          </p>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
          </p>
        </div>}
      </div>
    </>
    }
  </div>
}

export default observer(ProductDetail);