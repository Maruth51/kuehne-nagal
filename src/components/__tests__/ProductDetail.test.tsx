import { render, screen } from "@testing-library/react";
import ProductDetail from "components/ProductDetail";
import { StoreProvider } from "core/providers/StoreProvider";
import AppStore from "core/state/AppStore";
import { IProduct } from "core/types/Product.types";

const productTestData: IProduct = {
  "productName": "Foxit software PhantomPDF Standard",	
  "tags":[ "PDF", "Change", "Create", "Maintenance", "Business, FoxIT" ],
  "category": "Daily Business",
  "manufacturerUrl": "https://www.foxitsoftware.com/de/pdf-editor",
  "description": 
  [
    "PhantomPDF provides powerful PDF Editor capabilities to allow authors to update their documents themselves.",
    "Standard - Simple interface and limited functionality."
  ],
  "option1": "1 Year Maintenance",
  "option2": "Without Maintenance"
}

describe('Product detail component', () => {
  let store: AppStore;
  beforeEach(() => {
    store = new AppStore()
  })
  test('Not rendered on initial', () => {
    render(<StoreProvider store={store}>
            <ProductDetail />
          </StoreProvider>)
    const productDetailTitlElem = screen.queryByText(/Product Detail/i)
    expect(productDetailTitlElem).toBeFalsy()
  })

  test('Rendered upon selecting product', () => {
    store.selectedProduct = productTestData
    render(<StoreProvider store={store}>
      <ProductDetail />
    </StoreProvider>)
    const productDetailElem= screen.queryByText(/Product Detail/i)
    expect(productDetailElem).toBeInTheDocument()
  })

  test('Product name displayed', () => {
    store.selectedProduct = productTestData
    render(<StoreProvider store={store}>
      <ProductDetail />
    </StoreProvider>)
    const productDetailElem= screen.getByTestId('product-detail')
    expect(productDetailElem).toMatchSnapshot('ProductDetailComponent')
  })
})