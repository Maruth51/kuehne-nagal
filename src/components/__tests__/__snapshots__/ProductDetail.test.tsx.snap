// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product detail component Product name displayed: ProductDetailComponent 1`] = `
<div
  class="container"
  data-testid="product-detail"
>
  <div
    class="head"
  >
    Product Details
  </div>
  <div
    class="productDetail"
  >
    <span>
      Foxit software PhantomPDF Standard
       
    </span>
    <div
      class="tags"
    >
      <span
        class="tag"
      >
        PDF
      </span>
      <span
        class="tag"
      >
        Change
      </span>
      <span
        class="tag"
      >
        Create
      </span>
      <span
        class="tag"
      >
        Maintenance
      </span>
      <span
        class="tag"
      >
        Business, FoxIT
      </span>
    </div>
    <a
      class="actionBtn"
      href="https://www.foxitsoftware.com/de/pdf-editor"
      rel="noreferrer"
      target="_blank"
      type="button"
    >
      Go to Manufacturer
    </a>
    <p
      class="description"
    >
      PhantomPDF provides powerful PDF Editor capabilities to allow authors to update their documents themselves.
    </p>
    <div
      class="option"
    >
      <input
        id="option1"
        name="maintanance-option"
        type="radio"
      />
      <label
        for="option1"
      >
         
        1 Year Maintenance
      </label>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
      </p>
    </div>
    <div
      class="option"
    >
      <p>
        <input
          id="option2"
          name="maintanance-option"
          type="radio"
        />
        <label
          for="option2"
        >
           
          Without Maintenance
        </label>
      </p>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
      </p>
    </div>
  </div>
</div>
`;
